src/Emulator/Peripherals/Peripherals/MTD/OpenTitan_FlashController.cs

```diff
@@ -0,0 +1,1185 @@
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/IsOperationAllowed: ret = {0}",
+                ret);
+
+            if(!ret){
+                this.Log(
+                    LogLevel.Warning, "OpenTitan_FlashController/IsOperationAllowed: " +
+                    "Operation not allowed!");
+            }
+
+            return ret;
+        }
+
+        private bool IsOperationAllowedInMpRegion(OperationType opType, long mpRegisterOffset){
+            bool ret = false;
+
+            uint MP_REGION_CFG_N_VALUE = registersCollection.Read(mpRegisterOffset);
+            uint MP_REGION_CFG_N_RD_EN_FLAG = (0x1 << 1) & MP_REGION_CFG_N_VALUE;
+            uint MP_REGION_CFG_N_PROG_EN_FLAG = (0x1 << 2) & MP_REGION_CFG_N_VALUE;
+            uint MP_REGION_CFG_N_ERASE_EN_FLAG = (0x1 << 3) & MP_REGION_CFG_N_VALUE;
+
+            if(opType == OperationType.READ_DATA){
+                ret = MP_REGION_CFG_N_RD_EN_FLAG != 0;
+            }else if(opType == OperationType.PROGRAM_DATA){
+                ret = MP_REGION_CFG_N_PROG_EN_FLAG != 0;
+            }else if(opType == OperationType.ERASE_DATA_PAGE){
+                ret = MP_REGION_CFG_N_ERASE_EN_FLAG != 0;
+            }
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/IsOperationAllowedInMpRegion: ret = {0}",
+                ret);
+
+            if(!ret){
+                this.Log(
+                    LogLevel.Info, "OpenTitan_FlashController/IsOperationAllowedInMpRegion: " +
+                    "Operation not allowed!");
+            }
+
+            return ret;
+        }
+
+        private bool MpRegionRegisterAppliesToOperation(long mpRegisterOffset, long operationAddress){
+            uint MP_REGION_CFG_N_VALUE = registersCollection.Read(mpRegisterOffset);
+            uint MP_REGION_CFG_N_EN = MP_REGION_CFG_N_VALUE & 0x1;
+            uint MP_REGION_CFG_N_BASE = (0x1FF) & (MP_REGION_CFG_N_VALUE >> 8);
+            uint MP_REGION_CFG_N_SIZE = (0x3FF) & (MP_REGION_CFG_N_VALUE >> 20);
+
+            long regionStart = FLASH_MEM_BASE_ADDR + (MP_REGION_CFG_N_BASE * FLASH_WORDS_PER_PAGE * FLASH_WORD_SZ);
+            long regionEnd = regionStart + (MP_REGION_CFG_N_SIZE * FLASH_WORDS_PER_PAGE * FLASH_WORD_SZ);
+
+            if(MP_REGION_CFG_N_EN != 0x0
+                && operationAddress >= regionStart
+                && operationAddress < regionEnd){ // the bus is 4 bytes wide. Should that effect the border logic?
+                return true;
+            }
+
+            return false;
+        }
+
+        private const uint FLASH_WORDS_PER_PAGE = 128;
+        private const uint FLASH_WORD_SZ = 8;
+        private const uint FLASH_PAGES_PER_BANK = 256;
+        private const long FLASH_MEM_BASE_ADDR = 0x20000000;
+        private readonly DoubleWordRegisterCollection registersCollection;
+        public long Size => 0x1000;
+        private readonly MappedMemory flash;
+        private MappedMemory infoFlash;
```

Comment:
I think this can be `readonly` as well. We try to mark all possible fields as `readonly`.